server/src/main/java/org/elasticsearch/common/io/stream/RecyclerBytesStreamOutput.java

```diff
@@ -225,21 +224,24 @@ private void ensureCapacity(int bytesNeeded) {
     }
 
     private void ensureCapacityFromPosition(long newPosition) {
-        // Integer.MAX_VALUE is not a multiple of the page size so we can only allocate the largest multiple of the pagesize that is less
-        // than Integer.MAX_VALUE
-        if (newPosition > Integer.MAX_VALUE - (Integer.MAX_VALUE % pageSize)) {
+        final long desiredPageCount = (newPosition + pageSize - 1) / pageSize; // ceil(minTargetSize/pageSize)
+        final long pagesToAdd = desiredPageCount - pages.size();
+        if (pagesToAdd <= 0) {
+            return;
+        }
+        if (desiredPageCount * pageSize > Integer.MAX_VALUE) {
```

Comment:
I guess this still could be checked in the beginning using `newPosition` only? This way we would not need to do any computation in case of too high value for `newPosition`

Reply:
It needs to be rounded up to the nearest `pageSize` in any case so I think this is not unreasonable.

I wonder tho whether we lose performance by not checking `newPosition > currentCapacity` first. I expect these things see a lot of small writes, each of which hits this method, and a simple integer comparison could turn out cheaper than what we do here.

Reply:
Oh wait writes mostly go via `ensureCapacity` which avoids this. Nvm then.